File: python_sdk/src/reality_capture/specifications/water_constraints.py
from pydantic import BaseModel, Field
from typing import Optional
from enum import Enum


class WaterConstraintsInputs(BaseModel):
    scene: str = Field(description="Reality data ID (cloud) or local path (on-premise) of ContextScene")
    modeling_reference: str = Field(alias="modelingReference", description="Reality data ID (cloud) or local path (on-premise) of Modeling Reference")
    crs_data: Optional[str] = Field(default=None,
                                    description="Path to CRS data file (bucket path in cloud, local path on-premise).",
                                    alias="crsData")


class WaterConstraintsOptions(BaseModel):
    force_horizontal: Optional[bool] = Field(None, alias="forceHorizontal",
                                             description="Force constraints to be horizontal")


class WaterConstraintsOutputsCreate(Enum):
    CONSTRAINTS = "constraints"


class WaterConstraintsOutputs(BaseModel):
    constraints: str = Field(description="Path to output constraints file (bucket path in cloud, local path on-premise)")


class WaterConstraintsSpecificationsCreate(BaseModel):
    inputs: WaterConstraintsInputs = Field(description="Inputs")
    outputs: list[WaterConstraintsOutputsCreate] = Field(description="Outputs")
    options: Optional[WaterConstraintsOptions] = Field(None, description="Options")


class WaterConstraintsSpecifications(BaseModel):
    inputs: WaterConstraintsInputs = Field(description="Inputs")
    outputs: WaterConstraintsOutputs = Field(description="Outputs")
    options: Optional[WaterConstraintsOptions] = Field(None, description="Options")


class WaterConstraintsCost(BaseModel):
    gpix: float = Field(description="Number of GigaPixels in the overall inputs.", ge=0)

File: typescript/packages/reality-capture/src/specifications/water_constraints.ts
import { z } from "zod";

export enum WaterConstraintsOutputsCreate {
  CONSTRAINTS = "constraints",
}

export const WaterConstraintsInputsSchema = z.object({
    scene: z.string().describe("Reality data id of ContextScene"),
    modelingReference: z.string().describe("Reality data id of Modeling Reference"),
});
export type WaterConstraintsInputs = z.infer<typeof WaterConstraintsInputsSchema>;

export const WaterConstraintsOptionsSchema = z.object({
    forceHorizontal: z
        .boolean()
        .optional()
        .describe("Force constraints to be horizontal"),
});
export type WaterConstraintsOptions = z.infer<typeof WaterConstraintsOptionsSchema>;

export const WaterConstraintsOutputsSchema = z.object({
    constraints: z
        .string()
        .regex(/^bkt:.+/, { message: "Must match ^bkt:.+" })
        .describe("Path in the bucket of output constraints"),
});
export type WaterConstraintsOutputs = z.infer<typeof WaterConstraintsOutputsSchema>;

export const WaterConstraintsSpecificationsCreateSchema = z.object({
    inputs: WaterConstraintsInputsSchema.describe("Inputs"),
    outputs: z
        .array(z.nativeEnum(WaterConstraintsOutputsCreate))
        .describe("Outputs"),
    options: WaterConstraintsOptionsSchema.optional().describe("Options"),
});
export type WaterConstraintsSpecificationsCreate = z.infer<typeof WaterConstraintsSpecificationsCreateSchema>;

export const WaterConstraintsSpecificationsSchema = z.object({
    inputs: WaterConstraintsInputsSchema.describe("Inputs"),
    outputs: WaterConstraintsOutputsSchema.describe("Outputs"),
    options: WaterConstraintsOptionsSchema.optional().describe("Options"),
});
export type WaterConstraintsSpecifications = z.infer<typeof WaterConstraintsSpecificationsSchema>;

export const WaterConstraintsCostSchema = z.object({
    gpix: z
        .number()
        .min(0, { message: "Must be greater than or equal to 0" })
        .describe("Number of GigaPixels in the overall inputs."),
});
export type WaterConstraintsCost = z.infer<typeof WaterConstraintsCostSchema>;